trieste/acquisition/rule.py

```diff
@@ -200,14 +202,144 @@ def acquire(
+        :param optimizer: The optimizer with which to optimize the acquisition function built by
+            ``builder``. This should *maximize* the acquisition function, and must be compatible
+            with the global search space. Defaults to
+            :func:`~trieste.acquisition.optimizer.automatic_optimizer_selector`.
+        :param num_query_points: The number of points to acquire.
+        """
+        if num_query_points <= 0:
+            raise ValueError(
+                f"Number of query points must be greater than 0, got {num_query_points}"
+            )
+
+        if builder is None:
+            raise ValueError("Please specify an acquisition builder")
+
+        if not isinstance(
+            builder, (GreedyAcquisitionFunctionBuilder, SingleModelGreedyAcquisitionBuilder)
+        ):
+            raise NotImplementedError(
+                f"""Only greedy acquisition strategies are supported
+                    by AsyncEGO, got {type(builder)}"""
+            )
+
+        if optimizer is None:
+            optimizer = automatic_optimizer_selector
+
+        if isinstance(builder, SingleModelGreedyAcquisitionBuilder):
+            builder = builder.using(OBJECTIVE)
+
+        self._builder: GreedyAcquisitionFunctionBuilder = builder
+        self._optimizer = optimizer
+        self._num_query_points = num_query_points
+        self._acquisition_function: Optional[AcquisitionFunction] = None
+
+    def __repr__(self) -> str:
+        """"""
+        return f"""AsyncEGO(
+        {self._builder!r},
+        {self._optimizer!r},
+        {self._num_query_points!r})"""
+
+    def acquire(
+        self,
+        search_space: SP_contra,
+        datasets: Mapping[str, Dataset],
+        models: Mapping[str, ProbabilisticModel],
+    ) -> types.State[S | None, TensorType]:
+        """
+        Constructs a function that, given ``AsyncEGO.State``,
+        returns a new state object and points to evaluate.
+        The state object contains currently known pending points,
+        that is points that were requested for evaluation,
+        but observation for which was not received yet.
+        To keep them up to date, pending points are compared against the given dataset,
+        and whatever points are in the dataset are deleted.
+        Then the current batch is generated by calling the acquisition function,
+        and all points in the batch are added to the known pending points.
+
+        :param search_space: The local acquisition search space for *this step*.
+        :param datasets: The known observer query points and observations.
+        :param models: The models of the specified ``datasets``.
+        :return: A function that constructs the next acquisition state and the recommended query
+            points from the previous acquisition state.
+        """
+
+        def state_func(state: AsyncEGO.State | None) -> tuple[AsyncEGO.State | None, TensorType]:
```

Comment:
We normally call the state functions `go` for some reason (e.g. see TrustRegion).

```suggestion
        def go(state: AsyncEGO.State | None) -> tuple[AsyncEGO.State | None, TensorType]:
```

Reply:
Since there is only one such instance so far, I dare challenging it! `go` is rather cryptic, I find `state_func` clearer. If anything, probably we shall update TrustRegion?

Reply:
If we make this change, lets make it consistent across states

Reply:
i made the change in TrustRegion too